### ISSUE TITLE
bad error handling in ServerTimeoutsTest.testBlockingReadWithMinimumDataRateBelowLimit 

### ISSUE THREAD

**Issue by gregw** [eclipse/jetty.project#1076]
```
2016-11-07 08:55:37.467:INFO::main: Logging initialized @276ms to org.eclipse.jetty.util.log.StdErrLog
Running org.eclipse.jetty.http.client.ServerTimeoutsTest.testBlockingReadWithMinimumDataRateBelowLimit[transport: HTTP]()
2016-11-07 08:55:37.528:INFO:oejs.Server:main: jetty-9.4.z-SNAPSHOT
2016-11-07 08:55:37.604:INFO:oejs.AbstractConnector:main: Started ServerConnector@161cd475{HTTP/1.1,[http/1.1]}{0.0.0.0:39375}
2016-11-07 08:55:37.605:INFO:oejs.Server:main: Started @413ms
2016-11-07 08:55:37.641:INFO:oejus.SslContextFactory:main: x509=X509@4c3e4790(mykey,h=[],w=[]) for SslContextFactory@38cccef(file:///home/gregw/src/jetty-9.4.x/tests/test-http-client-transport/src/test/resources/keystore.jks,file:///home/gregw/src/jetty-9.4.x/tests/test-http-client-transport/src/test/resources/truststore.jks)
2016-11-07 08:55:39.858:WARN:oejs.HttpChannel:server-16: / org.eclipse.jetty.http.BadMessageException: 408: Request data rate < 20 B/s
2016-11-07 08:55:39.858:WARN:oejs.HttpChannel:server-16: / org.eclipse.jetty.http.BadMessageException: 408: Request data rate < 20 B/s
2016-11-07 08:55:39.858:WARN:oejs.HttpChannel:server-16: 
java.lang.IllegalStateException: Error already set
	at org.eclipse.jetty.server.HttpChannelState.onError(HttpChannelState.java:723)
	at org.eclipse.jetty.server.HttpChannel.handleException(HttpChannel.java:494)
	at org.eclipse.jetty.server.HttpChannelOverHttp.handleException(HttpChannelOverHttp.java:446)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:428)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:265)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:97)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$1.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)
Caused by: 
org.eclipse.jetty.http.BadMessageException: 408: Request data rate < 20 B/s
	at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:279)
	at org.eclipse.jetty.server.HttpInput.read(HttpInput.java:250)
	at org.eclipse.jetty.http.client.ServerTimeoutsTest$5.handle(ServerTimeoutsTest.java:509)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:564)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:318)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:265)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:97)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$1.run(QueuedThreadPool.java:590)
	at java.lang.Thread.run(Thread.java:745)
2016-11-07 08:55:40.867:INFO:oejs.AbstractConnector:main: Stopped ServerConnector@161cd475{HTTP/1.1,[http/1.1]}{0.0.0.0:0}
```

**Comment by gregw** [comment]
The issue appears to be that the handler correctly throws the `BadMessageException: 408` exception, which is caught by the HttpChannel.  But rather than being used to generate the 408 response an ERROR dispatch is done back to the original test handler, which tries to read again and we get the double error.

The test is self can easily be fixed by checking for ERROR dispatch, but I think that is covering up the problem.  Should we be doing an error dispatch in the first place or should we just be calling sendError?

**Comment by gregw** [comment]
In 9.4, we handle ERROR dispatches in the ContextHandler, where is calls sendError there.  The reason was to ensure that sendError was called correctly scoped by the context.

If we are running without a context handler, then either we have to let the handlers deal with error dispatches themselves OR we call sendError for them.  As it is hard to detect if there is a ContextHandler in the tree, I think we need to do the former (which is closer to the behaviour < 9.4 anyway).

**Comment by gregw** [comment]
Ah we had already thought of this in AbstractHandler, which has the code to generate the error page.  In this case the problem was the test handler overrode handle instead of doHandle!

**Comment by sbordet** [comment]
@gregw is this in 9.4 ? 

If so, and we require to change the implementations of `AbstractHandler.handle()` to override `AbstractHandler.doHandle()` instead, we are breaking 100% of users out there that implemented 9.3's `AbstractHandler.handle()`.

Unless I miss something, 9.4 should change to be backwards compatible.

I don't think it's right to fix the code by changing the override to `doHandle()`,

**Comment by gregw** [comment]
closing in favour of meta issue #1077
